torch/optim/adam.py

```diff
@@ -278,7 +284,10 @@ def _single_tensor_adam(params: List[Tensor],
 
             if amsgrad:
                 # Maintains the maximum of all 2nd moment running avg. till now
-                torch.maximum(max_exp_avg_sqs[i], exp_avg_sq, out=max_exp_avg_sqs[i])
+                if differentiable:
+                    max_exp_avg_sqs[i][:] = torch.maximum(max_exp_avg_sqs[i].clone(), exp_avg_sq)
```

Comment:
nit: `max_exp_avg_sqs[i].copy_(torch.maximum(max_exp_avg_sqs[i], exp_avg_sq))`

Also the perf hit compared to the out= call below is really minor. I think we can unconditionally use this new version for the sake of simplicity (same below). cc @jbschlosser what do you think?

Reply:
One thing here is that we need to clone `max_exp_avg_sqs` for this to work ...
Wonder if its ok to keep this

Reply:
Ho right maximum saves both its inputs.